navigation_and_routing/lib/src/auth/auth.dart

```diff
@@ -0,0 +1,53 @@
+// Copyright 2021, the Flutter project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+
+/// A mock authentication service
+class BookstoreAuth extends ChangeNotifier {
+  bool _signedIn = false;
+
+  bool get signedIn {
+    return _signedIn;
+  }
+
+  Future<void> signOut() async {
+    await Future<void>.delayed(const Duration(milliseconds: 200));
+    // Sign out.
+    _signedIn = false;
+    notifyListeners();
+  }
+
+  Future<bool> signIn(String username, String password) async {
+    await Future<void>.delayed(const Duration(milliseconds: 200));
+
+    // Sign in. Allow any password.
+    _signedIn = true;
+    notifyListeners();
+    return _signedIn;
+  }
+
+  @override
+  bool operator ==(Object other) {
+    return other is BookstoreAuth && other._signedIn == _signedIn;
+  }
+
+  @override
+  int get hashCode => _signedIn.hashCode;
+}
+
+class BookstoreAuthScope extends InheritedNotifier<BookstoreAuth> {
+  const BookstoreAuthScope({
+    required BookstoreAuth notifier,
+    required Widget child,
+    Key? key,
+  }) : super(key: key, notifier: notifier, child: child);
+
+  static BookstoreAuth? of(BuildContext context) {
```

Comment:
I would consider making the return type here non-nullable and make this throw a meaningful error if no BookstoreAuth can be found. This avoids the `!` when you call `BookstoreAuthScope.of` and also avoids the mental overhead there of wondering if this can ever be null.